packages/firestore/src/util/async_queue.ts

```diff
@@ -118,11 +154,42 @@ export class AsyncQueue {
           // and return the rejected Promise.
           throw error;
         })
-        .then(() => {
+        .then(result => {
           this.operationInProgress = false;
+          return result;
         });
     });
-    return this.tail as AnyDuringMigration;
+    this.tail = newTail;
+    return newTail;
+  }
+
+  /**
+   * Schedules an operation to be run on the AsyncQueue once the specified
+   * `delayMs` has elapsed. The returned DelayedOperationResult can be
+   * used to cancel the operation prior to its running.
+   */
+  scheduleWithDelay<T>(
+    op: () => Promise<T>,
+    delayMs: number
+  ): DelayedOperationResult<T> {
+    this.verifyNotFailed();
+
+    const delayedOp = DelayedOperation.createAndSchedule(this, op, delayMs);
+    const index = this.delayedOperations.push(delayedOp);
+
+    delayedOp.promise.catch(err => {}).then(() => {
+      this.delayedOperations.slice(index, 1);
```

Comment:
This doesn't seem right. The index might have been changed if another promise got fulfilled before this one.

Reply:
Yeesh. :-( Good catch, thank you!